# Math Riddle: Johnny the (Old Fashioned) Painter

<b>Johnny has been hired to paint the numbers 1 through 100.</b><br>
<i> Johnny is old fashioned because he paints numbers below the power of 2 <b>and</b> because we'll be solving using a flat list iterator with low generalizability.</i><br>
<b>   How many times will he paint the letter "2"?</b><br>

<hr>
<br>
<center> <h3> Johnny!</h3> </center>
<br>
<img style="-webkit-user-select: none;cursor: zoom-in;" src="https://image1.masterfile.com/getImage/846-08140099em-1930s-1940s-man-in-white-overalls-by-step-ladder-zinc-pail-bucket.jpg" width="360" height="283">
<br>
<center> <i> He's a great painter and has a simple yet fullfilling life.</i> </center>
<br>


**We'll solve using list comprehensions!**

   1. Create a list of numbers through 100.
   
   2. Flatten that list to strings of each digit.
   
   3. Define a function that iterates over your list and counts the number
   
   4. Ask for user input and push out a result!
   

In [53]:
# Step 1 – Create a list of numbers

z=[]
for i in range(100):
    z.append(str(i))
    
# Print your stuff:
print('Your list "z" has a length of: {0}\n'.format(len(z)))
print(z)

####################################

# Step 2 – Flatten that list.

z_flat = [y for x in z for y in x]

# Print your stuff:
print('\n\nYour flat list "z_flat" has a length of: {0}\n'.format(len(z_flat)))
print(z_flat)

####################################

# Step 3 – Define a function that iterates and counts

def item_counter(string_to_look_for = '2', list_to_look_in = z_flat):
    item_count = 0
    for item in list_to_look_in:
        if item == string_to_look_for:
            item_count +=1
    return item_count

# Print your stuff:
print('\n\nYour function "item counter" has been added successfully.')

def item_counter_2(string_to_look_for = '2', list_to_look_in = z):
    item_count = 0    
    for item in list_to_look_in:
        if string_to_look_for in item:
            item_count +=1
    return item_count
item_counter_2()

# Print your stuff:
print('\n\nYour function "item counter" has been added successfully.')


Your list "z" has a length of: 100

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45', '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58', '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71', '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84', '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97', '98', '99']


Your flat list "z_flat" has a length of: 190

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '1', '0', '1', '1', '1', '2', '1', '3', '1', '4', '1', '5', '1', '6', '1', '7', '1', '8', '1', '9', '2', '0', '2', '1', '2', '2', '2', '3', '2', '4', '2', '5', '2', '6', '2', '7', '2', '8', '2', '9', '3', '0', '3', '1', '3', '2', '3', '3', '3', '4', '3', '5', '3', '6', '3',

In [54]:
# Step 4 – Ask for user input and execute

num_look_for = input('Choose a number you want to count in John\'s painting job:\n')
if int(num_look_for) >9:
    list_use = z
else:
    list_use = z_flat
    
    
####################################

# Print your stuff: (The results)

print('\nJohnny paints {0} a total of {1} times.\n'.format(num_look_for,
                                                       item_counter(num_look_for,list_use)))

Choose a number you want to count in John's painting job:
9

Johnny paints 9 a total of 20 times.



# How about a faster way to do it:

In [56]:
# Step 4 – Ask for user input and execute

num_look_for = input('Choose a number you want to count in Johnny\'s painting job:\n')
if int(num_look_for) >9:
    count_value = 3
else:
    count_value = ''.join(str(i) for i in range(1,101)).count(num_look_for)
    
    
####################################

# Print your stuff: (The results)

print('\nJohnny paints {0} a total of {1} times.\n'.format(num_look_for,count_value))


Choose a number you want to count in Johnny's painting job:
9

Johnny paints 9 a total of 20 times.



# Janie
<b>Janie has been hired to paint the numbers 1 through 1,000,000.</b><br>
    <i> She's got a very hard job, especially compared to Johnny from the good old days. But Janie loves what she does. Except, she for counting how many times she'll paint a certain number! This part of the job frsutrates her to no end. Let's help her?</i><br>
   <b>How many times will she paint "19"?</b><br>

<hr>

<img style="-webkit-user-select: none;cursor: zoom-in;" src="https://media.istockphoto.com/photos/paintercollection-in-my-gallery-picture-id524298601?k=6&amp;m=524298601&amp;s=612x612&amp;w=0&amp;h=8ltm2aOOynE9nZ1_o_LmLqUSyaojt0DKlfVXbPIrTXI=" width="360" height="456">
<br>
<center> <i> She's been painting for a while now... <br>
    It's not the paint which wears her down... it's the counting.<br>
    Fortunately, our programming skills are going to save her! </i> </center>
<br>

In [57]:
def fast_item_counter(string_to_look_for = '19',
                      list_to_look_in = range(1,1000001)):
    item_count = 0 
    # for every number in the list:
    for i in list_to_look_in:
        # if '19' is in that nunber:
        if string_to_look_for in str(i):
            # add to my count...
            item_count +=1 
            # Take the numbers after the instance of 19
            # For example 1192 becomes 2)
            item_split = str(i).split(string_to_look_for,1)[1]
            # While '19' is in the split string:
            while string_to_look_for in item_split: 
                print(i) # print that number...
                print(item_split) # print the split too...
                if string_to_look_for in item_split:
                    # add to my count
                    item_count +=1
                    # Then split again!
                    item_split = item_split.split(string_to_look_for,1)[1]
                
    return item_count

<b><i>My numbers get split accordingly:</b></i> <br><br>
**1919:** 19<br>
**11919:** 19 <br>
**19019:** 019 <br>
**19119:** 119 <br>
**19190:** 190 <br>
**19191:** 191 <br>
**19192:** 192

### We want to test out our function:

We'll cycle through the two numbers: 191,918 and 191,919.

* Our function should split the first number into: 1918 <br>
* Our function should split the second number into: 1919 and then 19<br>
* **Our total count should be: 5**


In [32]:
z = fast_item_counter(num_look_for,range(191918,191920))
print('\nThe number "19" was found a total of {0} times'.format(z))

191918
1918
191919
1919
191919
19

The number "19" was found a total of 5 times


Indeed, the function does as it should:
* 191,918 was split once and counted twice
* 191,919 was split twice and counted three times
* **Total result was 5!**



## Now that the testing is over, remove any "print" fields unnecessary to core execution of the function:

In [58]:
def fast_item_counter(string_to_look_for = '19',
                      list_to_look_in = range(1,1000001)):
    item_count = 0 
    # for every number in the list:
    for i in list_to_look_in:
        # if '19' is in that nunber:
        if string_to_look_for in str(i):
            # add to my count...
            item_count +=1 
            # Take the numbers after the instance of 19
            # For example 1192 becomes 2)
            item_split = str(i).split(string_to_look_for,1)[1]
            # While '19' is in the split string:
            while string_to_look_for in item_split: 
                # No printing, only good hard results
                if string_to_look_for in item_split:
                    # add to my count
                    item_count +=1
                    # Then split again!
                    item_split = item_split.split(string_to_look_for,1)[1]
                
    return item_count

In [51]:
num_look_for = input('Choose a number you want to count in Janie\'s painting job:\n')
#range_to_look_in = [int(x) for x in input('Choose the range you want to use:\n').split(',')]
# range(range_to_look_in[0],range_to_look_in[1]
range_to_use = range(1,1000001)
count_value = fast_item_counter(num_look_for,range_to_use)
    
    
####################################

# Print your stuff: (The results)
print('\nJanie paints {0} a total of {1} times.\nWhen she paints {2} to {3}.'.format(num_look_for,count_value,range_to_use[0], range_to_use[-1]))


Choose a number you want to count in Janie's painting job:
19

Janie paints 19 a total of 50000 times.
When she paints 1 to 1000000.
